tests/failpoints/cases/test_conf_change.rs

```diff
@@ -204,3 +209,79 @@ fn test_stale_peer_cache() {
     fail::cfg("stale_peer_cache_2", "return").unwrap();
     cluster.must_put(b"k2", b"v2");
 }
+
+// The test is for this situation:
+// suppose there are 3 peers (1, 2, and 3) in a Raft group, and then
+// 1. propose to add peer 4 on the current leader 1;
+// 2. leader 1 appends entries to peer 3, and peer 3 applys them;
+// 3. leadersip is transfered to peer 2, and a new proposal to remove peer 4 is proposed;
+// 4. peer 2 sends a snapshot with latest configuration [1, 2, 3] to peer 3;
+// 5. peer 3 restores the snapshot into memory;
+// 6. then peer 3 calling `Raft::apply_conf_change` to add peer 4;
+// 7. so the disk configuration `[1, 2, 3]` is different from memory configuration `[1, 2, 3, 4]`.
+#[test]
+fn test_redundant_conf_change_by_snapshot() {
+    let _guard = crate::setup();
+
+    let mut cluster = new_node_cluster(0, 4);
+    cluster.cfg.raft_store.raft_log_gc_count_limit = 5;
+    cluster.cfg.raft_store.merge_max_log_gap = 4;
+    cluster.cfg.raft_store.raft_log_gc_tick_interval = ReadableDuration::millis(20);
+    // cluster.cfg.raft_store.hibernate_regions = false;
+
+    let pd_client = Arc::clone(&cluster.pd_client);
+    pd_client.disable_default_operator();
+
+    // Add voter 2 and learner 3.
+    cluster.run_conf_change();
+    pd_client.must_add_peer(1, new_peer(2, 2));
+    pd_client.must_add_peer(1, new_peer(3, 3));
+    cluster.must_put(b"k1", b"v1");
+    must_get_equal(&cluster.get_engine(2), b"k1", b"v1");
+    must_get_equal(&cluster.get_engine(3), b"k1", b"v1");
+
+    fail::cfg("apply_on_conf_change_3_1", "pause").unwrap();
+    cluster.pd_client.must_add_peer(1, new_peer(4, 4));
+
+    let filter = Box::new(
+        RegionPacketFilter::new(1, 3)
+            .direction(Direction::Recv)
+            .msg_type(MessageType::MsgAppend),
+    );
+    cluster.sim.wl().add_recv_filter(3, filter);
+
+    // Transfer leader to peer 2, and append more entries to compact raft logs.
+    cluster.must_transfer_leader(1, new_peer(2, 2));
+    cluster.pd_client.must_remove_peer(1, new_peer(4, 4));
+    (0..10).for_each(|_| cluster.must_put(b"k2", b"v2"));
+    sleep_ms(50);
+
+    // Clear filters on peer 3, so it can receive and restore a snapshot.
+    cluster.sim.wl().clear_recv_filters(3);
+    sleep_ms(100);
```

Comment:
Why sleep?

Reply:
Without sleep the snapshot may come after peer 3 calling `apply_conf_change`.